lib/CoreBot.js

```diff
@@ -40,13 +47,75 @@ function Botkit(configuration) {
     // define some middleware points where custom functions
     // can plug into key points of botkits process
     botkit.middleware = {
-        send: ware(),
-        receive: ware(),
         spawn: ware(),
-        heard: ware(),
+        ingest: ware(),
+        normalize: ware(),
+        categorize: ware(),
+        receive: ware(),
+        triggered: ware(), // triggered any event
+        heard: ware(), // best place for heavy i/o because fewer messages
         capture: ware(),
+        send: ware(),
+    };
+
+
+    botkit.ingest = function(bot, payload, source) {
+
+        // keep an unmodified copy of the message
+        payload.original_message = clone(payload);
+
+        payload._pipeline = {
+            stage: 'ingest',
+        };
+
+        // debug('ingest', payload);
+
+        botkit.middleware.ingest.run(bot, payload, source, function(err, bot, payload, source) {
+            botkit.normalize(bot, payload);
+        });
+    };
+
+    botkit.normalize = function(bot, payload) {
+        payload._pipeline.stage = 'normalize';
+        // debug('normalize', payload);
+        botkit.middleware.normalize.run(bot, payload, function(err, bot, message) {
+            if (!message.type) {
+                message.type = 'message_received';
+            }
+            debug('normalized', message);
+            botkit.categorize(bot, message);
+        });
     };
 
+    botkit.categorize = function(bot, message) {
+        // debug('categorize', message);
+        message._pipeline.stage = 'categorize';
+        botkit.middleware.categorize.run(bot, message, function(err, bot, message) {
+            botkit.receiveMessage(bot, message);
+        });
+    };
+
+    botkit.receiveMessage = function(bot, message) {
+        // debug('receive', message);
+        message._pipeline.stage = 'receive';
+        botkit.middleware.receive.run(bot, message, function(err, bot, message) {
+            if (err) {
+                botkit.log('ERROR IN RECEIVE MIDDLEWARE: ', err);
```

Comment:
Correct. If an error happens in any of the middlewares, the processing will stop.

It could be a syntax error in the middleware, or an actual error being thrown. You can call next(err), or you can actually throw an error in a middleware and it will trigger the middleware to STOP IMMEDIATELY and return the error.